crates/teloxide/src/dispatching/tracing.rs

```diff
@@ -0,0 +1,29 @@
+use super::{DpHandlerDescription, UpdateHandler};
+
+use dptree::{prelude::DependencyMap, HandlerDescription};
+use tracing::Instrument;
+
+/// Extension methods for adding `tracing` to [`UpdateHandler`].
+///
+/// This trait is only available if the `tracing` feature is enabled.
+pub trait UpdateHandlerTracingExt<E> {
+    /// Returns an `UpdateHandler` with tracing enabled.
+    fn with_tracing_span<F>(self, f: F) -> Self
+    where
+        F: Fn(DependencyMap) -> tracing::Span + Send + Sync + 'static;
```

Comment:
To use `Injectable` the bound should be something like this:
```rust
F: Injectable<DependencyMap, tracing::Span, FnArgs>
```

Where `FnArgs` is a newly added generic parameter. (btw you should probably import `tracing::Span`)

Then you can use it like this:
```rust
f.inject(&deps)()
```

Reply:
I tried implementing it with `Injectable` but the issue I'm having is that the `.inject()` returns a `Future` that I have to await in order to get `tracing::Span`, but `dptree::from_fn_with_description` doesn't expect an async closure.

For reference, this is the code I'm testing with:

```rust
pub trait UpdateHandlerTracingExt<E> {
    fn with_tracing_span<F, FnArgs>(self, f: F) -> Self
    where
        F: Injectable<DependencyMap, Span, FnArgs>;
}

impl<E: 'static> UpdateHandlerTracingExt<E> for UpdateHandler<E> {
    fn with_tracing_span<F, FnArgs>(self, f: F) -> Self
    where
        F: Injectable<DependencyMap, Span, FnArgs>,
    {
        let description = DpHandlerDescription::entry().merge_chain(self.description());

        dptree::from_fn_with_description(description, move |deps: DependencyMap, cont| async {
            let span = f.inject(&deps)().await;
            self.clone().execute(deps, cont).instrument(span)
        })
    }
}

```

Reply:
You can simply add an `.await` after the instrument and it should work:
```rust
self.clone().execute(deps, cont).instrument(span).await
```